x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -569,8 +570,18 @@ private void resolveInferences(
     }
 
     static PreAnalysisResult fieldNames(LogicalPlan parsed, Set<String> enrichPolicyMatchFields, PreAnalysisResult result) {
-        if (false == parsed.anyMatch(plan -> plan instanceof Aggregate || plan instanceof Project)) {
+        List<LogicalPlan> inlinestats = parsed.collect(InlineStats.class::isInstance);
+        Set<Aggregate> inlinestatsAggs = new HashSet<>();
+        for (var i : inlinestats) {
+            inlinestatsAggs.add(((InlineStats) i).aggregate());
+        }
```

Comment:
The required solution here looks correct but confusing; this is because we parse `INLINESTATS` as an `InlineStats` node containing an `Aggregate` node as child, so we don't know for any given `Aggregate` if it's a `STATS` or an `INLINESTATS`, and the two have very different semantics.

I think we should rather just parse `INLINESTATS` as a single plan node - this would prevent this complexity.

Maybe consider refactoring the `InlineStats` node to avoid adding complexity here, as the `fieldNames` method is already hard to work with. A low effort fix would be to still have the `InlineStats` wrap an `Aggregate`, but not as its child - the actual child would be the preceding command.

Reply:
Those are some good points (the use of `GeneratingPlan` and refactoring `InlineStats`), but I need more time to dig through these to prove these are valid changes to make. IMHO, the argument for simplifying what `fieldNames` is doing (looking at the aggregate inside an inlinestats) is not a strong one to warrant the refactoring. This change needs to be conceptually sound to make sense, ignoring the EsqlSession stuff.

Meaning, the conceptually sound argument **needs to drive** the refactoring and not the fact that `fieldNames` becomes more complex.